File: cpp_runtime/core/models/deepseek_v41.cpp
#include "models/deepseek_v41.h"

#include "nlohmann/json.hpp"

#include <algorithm>
#include <cmath>
#include <stdexcept>

namespace mfq::models::deepseek_v41 {
namespace {

using json = nlohmann::json;

const json& object_at(const json& root, const char* name) {
    const auto found = root.find(name);
    if (found == root.end() || !found->is_object()) {
        throw std::runtime_error(
            std::string("DeepSeek-V4.1 config requires object ") + name);
    }
    return *found;
}

std::int64_t positive(const json& value, const char* name) {
    const auto found = value.find(name);
    if (found == value.end() || !found->is_number_integer() ||
        found->get<std::int64_t>() <= 0) {
        throw std::runtime_error(
            std::string("DeepSeek-V4.1 config requires positive ") + name);
    }
    return found->get<std::int64_t>();
}

std::vector<std::int64_t> integers(
    const json& value,
    const char* name) {
    const auto found = value.find(name);
    if (found == value.end() || !found->is_array()) {
        throw std::runtime_error(
            std::string("DeepSeek-V4.1 config requires array ") + name);
    }
    std::vector<std::int64_t> result;
    result.reserve(found->size());
    for (const auto& item : *found) {
        if (!item.is_number_integer()) {
            throw std::runtime_error(
                std::string("DeepSeek-V4.1 config array is not integral: ") + name);
        }
        result.push_back(item.get<std::int64_t>());
    }
    return result;
}

std::vector<std::int64_t> eos_ids(const json& root) {
    const auto found = root.find("eos_token_id");
    if (found == root.end() || found->is_null()) return {};
    if (found->is_number_integer()) {
        return {found->get<std::int64_t>()};
    }
    if (!found->is_array()) {
        throw std::runtime_error(
            "DeepSeek-V4.1 eos_token_id must be an integer or array");
    }
    std::vector<std::int64_t> result;
    for (const auto& item : *found) {
        if (!item.is_number_integer()) {
            throw std::runtime_error(
                "DeepSeek-V4.1 eos_token_id array is not integral");
        }
        result.push_back(item.get<std::int64_t>());
    }
    return result;
}

bool contains(
    const std::vector<std::int64_t>& values,
    std::int64_t value) noexcept {
    return std::find(values.begin(), values.end(), value) != values.end();
}

void require_layers(
    const std::vector<std::int64_t>& values,
    std::int64_t layers,
    const char* description) {
    if (std::any_of(
            values.begin(), values.end(),
            [layers](std::int64_t value) {
                return value < 0 || value >= layers;
            })) {
        throw std::runtime_error(
            std::string("DeepSeek-V4.1 ") + description +
            " contains an invalid layer");
    }
}

} // namespace

Config Config::from_json(std::string_view payload) {
    json root;
    try {
        root = json::parse(payload.begin(), payload.end());
    } catch (const json::exception& error) {
        throw std::runtime_error(
            std::string("invalid DeepSeek-V4.1 config JSON: ") + error.what());
    }
    if (!root.is_object() ||
        root.value("model_type", std::string{}) != "deepseek_v41") {
        throw std::runtime_error(
            "DeepSeek-V4.1 runtime requires model_type=deepseek_v41");
    }
    const auto& text = object_at(root, "text_config");
    if (text.value("model_type", std::string{}) != "deepseek_v41_text") {
        throw std::runtime_error(
            "DeepSeek-V4.1 runtime requires text model_type=deepseek_v41_text");
    }
    const auto& vision = object_at(root, "vision_config");
    if (vision.value("model_type", std::string{}) != "deepseek_v41_vision") {
        throw std::runtime_error(
            "DeepSeek-V4.1 runtime requires its native vision tower");
    }
    const auto& quantization = object_at(root, "quantization_config");
    const auto blocks = integers(quantization, "weight_block_size");
    if (quantization.value("quant_method", std::string{}) != "fp8" ||
        blocks != std::vector<std::int64_t>{32, 32} ||
        quantization.value("scale_fmt", std::string{}) != "ue8m0" ||
        quantization.value("expert_dtype", std::string{}) != "fp4") {
        throw std::runtime_error(
            "DeepSeek-V4.1 runtime does not support this source encoding");
    }

    Config config;
    config.vocab = positive(text, "vocab_size");
    config.hidden = positive(text, "hidden_size");
    config.n_layers = positive(text, "num_hidden_layers");
    config.n_heads = positive(text, "num_attention_heads");
    config.n_kv_heads = positive(text, "num_key_value_heads");
    config.head_dim = positive(text, "head_dim");
    config.rope_head_dim = positive(text, "qk_rope_head_dim");
    config.q_lora_rank = positive(text, "q_lora_rank");
    config.o_lora_rank = positive(text, "o_lora_rank");
    config.o_groups = positive(text, "o_groups");
    config.moe_inter = positive(text, "moe_intermediate_size");
    config.n_experts = positive(text, "n_routed_experts");
    config.n_shared = positive(text, "n_shared_experts");
    config.top_k = positive(text, "num_experts_per_tok");
    config.rms_eps = text.value("rms_norm_eps", 1e-20);
    config.routed_scaling = text.value("routed_scaling_factor", 1.0);
    config.swiglu_limit = text.value("swiglu_limit", 0.0);
    config.norm_topk_prob = text.value("norm_topk_prob", true);
    config.sliding_window = positive(text, "sliding_window");
    config.max_position_embeddings = positive(text, "max_position_embeddings");
    config.rope_theta = text.value("rope_theta", 10'000.0);
    config.compress_rope_theta = text.value("compress_rope_theta", 160'000.0);
    config.compress_ratios = integers(text, "compress_ratios");
    const auto decoder = std::find(
        config.compress_ratios.begin(), config.compress_ratios.end(), 1);
    config.causal_encoder_layers = decoder == config.compress_ratios.end()
        ? 0
        : std::distance(config.compress_ratios.begin(), decoder);
    config.kv_source_layers = integers(text, "kv_source_layer_ids");
    config.index_source_layers = integers(text, "index_source_layer_ids");
    config.index_n_heads = positive(text, "index_n_heads");
    config.index_head_dim = positive(text, "index_head_dim");
    config.index_topk = positive(text, "index_topk");
    config.candidate_source_layer = text.value("candidate_source_layer_id", -1);
    config.candidate_topk_blocks = positive(text, "candidate_topk_blocks");
    config.candidate_block_size = positive(text, "candidate_block_size");
    config.hc_mult = positive(text, "hc_mult");
    config.hc_sinkhorn_iters = positive(text, "hc_sinkhorn_iters");
    config.hc_eps = text.value("hc_eps", 1e-6);
    config.engram_layer_ids = integers(text, "engram_layer_ids");
    config.engram_num_embeddings = integers(text, "engram_num_embeddings");
    config.engram_max_ngram_size = positive(text, "engram_max_ngram_size");
    config.engram_vocab_size = positive(text, "engram_vocab_size");
    config.engram_n_heads = positive(text, "engram_n_heads");
    config.engram_head_dim = positive(text, "engram_head_dim");
    config.engram_pad_token_id = text.value("engram_pad_token_id", 0);
    config.engram_compressed_vocab_size = positive(
        text, "engram_compressed_vocab_size");
    config.n_mtp_layers = positive(text, "num_nextn_predict_layers");
    config.dspark_block_size = positive(text, "dspark_block_size");
    config.dspark_noise_token_id = text.value("dspark_noise_token_id", 0);
    config.dspark_target_layer_ids = integers(text, "dspark_target_layer_ids");
    config.dspark_markov_rank = positive(text, "dspark_markov_rank");
    config.dspark_n_experts = positive(text, "dspark_n_routed_experts");
    config.dspark_top_k = positive(text, "dspark_num_experts_per_tok");
    config.image_token_id = root.value("image_token_id", -1);
    config.eos_token_ids = eos_ids(root);

    const auto& rope = object_at(text, "rope_scaling");
    if (rope.value("rope_type", std::string{}) != "yarn") {
        throw std::runtime_error("DeepSeek-V4.1 runtime requires YaRN RoPE");
    }
    config.rope_scaling.factor = rope.value("factor", 1.0);
    config.rope_scaling.beta_fast = rope.value("beta_fast", 32.0);
    config.rope_scaling.beta_slow = rope.value("beta_slow", 1.0);
    config.rope_scaling.original_max_position_embeddings =
        positive(rope, "original_max_position_embeddings");

    config.vision.n_layers = positive(vision, "num_hidden_layers");
    config.vision.hidden = positive(vision, "hidden_size");
    config.vision.n_heads = positive(vision, "num_attention_heads");
    config.vision.intermediate = positive(vision, "intermediate_size");
    config.vision.patch_size = positive(vision, "patch_size");
    config.vision.downsample_ratio = positive(vision, "downsample_ratio");
    config.vision.max_image_tokens = positive(vision, "max_image_tokens");
    config.vision.min_pixels = positive(vision, "min_pixels");
    config.vision.rope_theta = vision.value("rope_theta", 10'000.0);
    config.validate();
    return config;
}

void Config::validate() const {
    if (vocab <= 0 || hidden <= 0 || n_layers <= 0 || n_heads <= 0 ||
        n_kv_heads <= 0 || head_dim <= 0 || rope_head_dim <= 0 ||
        q_lora_rank <= 0 || o_lora_rank <= 0 || o_groups <= 0 ||
        moe_inter <= 0 || n_experts <= 0 || n_shared <= 0 || top_k <= 0 ||
        top_k > n_experts || n_heads % n_kv_heads || n_heads % o_groups ||
        rope_head_dim > head_dim || rope_head_dim % 2 || rms_eps <= 0.0 ||
        sliding_window <= 0 || max_position_embeddings <= 0 ||
        hc_mult <= 0 || hc_sinkhorn_iters <= 0 || hc_eps <= 0.0) {
        throw std::runtime_error("invalid DeepSeek-V4.1 model dimensions");
    }
    if (compress_ratios.size() !=
        static_cast<std::size_t>(n_layers + n_mtp_layers)) {
        throw std::runtime_error(
            "DeepSeek-V4.1 compression schedule has the wrong length");
    }
    const auto decoder = std::find(
        compress_ratios.begin(),
        compress_ratios.begin() + n_layers,
        1);
    if (decoder == compress_ratios.begin() + n_layers) {
        throw std::runtime_error("DeepSeek-V4.1 CED schedule has no decoder");
    }
    const auto derived_encoder = std::distance(compress_ratios.begin(), decoder);
    if (causal_encoder_layers != 0 && causal_encoder_layers != derived_encoder) {
        throw std::runtime_error("DeepSeek-V4.1 CED split is inconsistent");
    }
    for (std::int64_t layer = 0; layer < n_layers; ++layer) {
        const auto ratio = compress_ratios[static_cast<std::size_t>(layer)];
        if (ratio < 0 || ratio > 2 ||
            (layer >= derived_encoder && ratio != 1)) {
            throw std::runtime_error("DeepSeek-V4.1 CED compression ratio is invalid");
        }
    }
    for (std::int64_t layer = n_layers;
         layer < n_layers + n_mtp_layers; ++layer) {
        if (compress_ratios[static_cast<std::size_t>(layer)] != 0) {
            throw std::runtime_error(
                "DeepSeek-V4.1 DSpark compression ratio must be zero");
        }
    }
    require_layers(kv_source_layers, n_layers, "KV source schedule");
    require_layers(index_source_layers, n_layers, "index source schedule");
    require_layers(engram_layer_ids, n_layers, "Engram schedule");
    require_layers(dspark_target_layer_ids, n_layers, "DSpark target schedule");
    if (kv_source_layers.empty() ||
        !std::all_of(
            kv_source_layers.begin(), kv_source_layers.end(),
            [this](std::int64_t layer) { return is_index_source(layer); }) ||
        !is_index_source(candidate_source_layer) ||
        engram_layer_ids.size() != engram_num_embeddings.size() ||
        dspark_target_layer_ids.size() != static_cast<std::size_t>(n_mtp_layers)) {
        throw std::runtime_error("DeepSeek-V4.1 architecture schedules disagree");
    }
    std::int64_t active_ratio = 0;
    for (std::int64_t layer = 0; layer < n_layers; ++layer) {
        const auto ratio = compress_ratios[static_cast<std::size_t>(layer)];
        if (is_kv_source(layer)) {
            if (ratio <= 0) {
                throw std::runtime_error(
                    "DeepSeek-V4.1 KV source has no compressed stream");
            }
            active_ratio = ratio;
        }
        if (ratio > 0 && active_ratio != ratio) {
            throw std::runtime_error(
                "DeepSeek-V4.1 CSA2 consumer has no compatible KV source");
        }
    }
}

bool Config::has_engram(std::int64_t layer) const noexcept {
    return contains(engram_layer_ids, layer);
}

bool Config::is_kv_source(std::int64_t layer) const noexcept {
    return contains(kv_source_layers, layer);
}

bool Config::is_index_source(std::int64_t layer) const noexcept {
    return contains(index_source_layers, layer);
}

std::string TensorNames::layer(
    std::size_t index,
    std::string_view suffix) {
    return "model.block." + std::to_string(index) + "." + std::string(suffix);
}

std::string TensorNames::predictor(
    std::size_t index,
    std::string_view suffix) {
    return "predictor.stage." + std::to_string(index) + "." + std::string(suffix);
}

} // namespace mfq::models::deepseek_v41

File: mfq/architectures/deepseek_v41.py
"""Strict architecture contract for DeepSeek-V4.1-Flash.

DeepSeek-V4.1 is a CED/CSA2 model with Engram conditional memory,
Mega-mHC residual streams, native vision, and a three-stage DSpark head.  It
is deliberately represented independently from the older DeepSeek-V4 family;
only architecture-neutral storage and kernel primitives may be shared.
"""

from __future__ import annotations

from collections.abc import Mapping, Sequence
from dataclasses import dataclass


def _object(value: object, name: str) -> Mapping[str, object]:
    if not isinstance(value, Mapping):
        raise ValueError(f"{name} must be an object")
    return value


def _positive(value: object, name: str) -> int:
    result = int(value or 0)
    if result <= 0:
        raise ValueError(f"{name} must be positive")
    return result


def _integers(value: object, name: str) -> tuple[int, ...]:
    if not isinstance(value, Sequence) or isinstance(value, (str, bytes)):
        raise ValueError(f"{name} must be an array")
    return tuple(int(item) for item in value)


def _eos_ids(value: object) -> tuple[int, ...]:
    if value is None:
        return ()
    if isinstance(value, int):
        return (value,)
    return _integers(value, "eos_token_id")


@dataclass(frozen=True)
class DeepseekV41VisionConfig:
    model_type: str
    num_hidden_layers: int
    hidden_size: int
    num_attention_heads: int
    intermediate_size: int
    patch_size: int
    rope_theta: float
    downsample_ratio: int
    max_image_tokens: int
    min_pixels: int
    max_wh_ratio: float | None

    @classmethod
    def from_mapping(cls, raw: Mapping[str, object]) -> DeepseekV41VisionConfig:
        model_type = str(raw.get("model_type", ""))
        if model_type != "deepseek_v41_vision":
            raise ValueError(f"expected deepseek_v41_vision, got {model_type!r}")
        ratio = _positive(raw.get("downsample_ratio"), "vision.downsample_ratio")
        heads = _positive(raw.get("num_attention_heads"), "vision.num_attention_heads")
        hidden = _positive(raw.get("hidden_size"), "vision.hidden_size")
        if hidden % heads:
            raise ValueError("DeepSeek-V4.1 vision hidden size must divide its heads")
        max_wh_ratio = raw.get("max_wh_ratio")
        return cls(
            model_type=model_type,
            num_hidden_layers=_positive(
                raw.get("num_hidden_layers"), "vision.num_hidden_layers"
            ),
            hidden_size=hidden,
            num_attention_heads=heads,
            intermediate_size=_positive(
                raw.get("intermediate_size"), "vision.intermediate_size"
            ),
            patch_size=_positive(raw.get("patch_size"), "vision.patch_size"),
            rope_theta=float(raw.get("rope_theta", 10_000.0)),
            downsample_ratio=ratio,
            max_image_tokens=_positive(
                raw.get("max_image_tokens"), "vision.max_image_tokens"
            ),
            min_pixels=_positive(raw.get("min_pixels"), "vision.min_pixels"),
            max_wh_ratio=None if max_wh_ratio is None else float(max_wh_ratio),
        )


@dataclass(frozen=True)
class DeepseekV41Config:
    family: str
    model_type: str
    text_model_type: str
    vocab_size: int
    hidden_size: int
    moe_intermediate_size: int
    num_hidden_layers: int
    causal_encoder_layers: int
    num_attention_heads: int
    num_key_value_heads: int
    head_dim: int
    qk_rope_head_dim: int
    q_lora_rank: int
    o_lora_rank: int
    o_groups: int
    rms_norm_eps: float
    max_position_embeddings: int
    rope_theta: float
    rope_factor: float
    original_max_position_embeddings: int
    n_routed_experts: int
    n_shared_experts: int
    num_experts_per_tok: int
    scoring_func: str
    norm_topk_prob: bool
    routed_scaling_factor: float
    swiglu_limit: float
    sliding_window: int
    compress_ratios: tuple[int, ...]
    compress_rope_theta: float
    kv_source_layer_ids: tuple[int, ...]
    index_source_layer_ids: tuple[int, ...]
    index_n_heads: int
    index_head_dim: int
    index_topk: int
    candidate_source_layer_id: int
    candidate_topk_blocks: int
    candidate_block_size: int
    hc_mult: int
    hc_sinkhorn_iters: int
    hc_eps: float
    engram_layer_ids: tuple[int, ...]
    engram_num_embeddings: tuple[int, ...]
    engram_max_ngram_size: int
    engram_vocab_size: int
    engram_n_heads: int
    engram_head_dim: int
    engram_pad_token_id: int
    engram_compressed_vocab_size: int
    num_nextn_predict_layers: int
    dspark_block_size: int
    dspark_noise_token_id: int
    dspark_target_layer_ids: tuple[int, ...]
    dspark_markov_rank: int
    dspark_n_routed_experts: int
    dspark_num_experts_per_tok: int
    dense_weight_block_size: tuple[int, int]
    dense_scale_format: str
    expert_dtype: str
    image_token_id: int
    eos_token_ids: tuple[int, ...]
    vision: DeepseekV41VisionConfig

    @classmethod
    def from_hf_config(cls, outer: Mapping[str, object]) -> DeepseekV41Config:
        outer_type = str(outer.get("model_type", ""))
        text = _object(outer.get("text_config"), "text_config")
        text_type = str(text.get("model_type", ""))
        if outer_type != "deepseek_v41" or text_type != "deepseek_v41_text":
            raise ValueError(
                f"expected deepseek_v41/deepseek_v41_text, got "
                f"{outer_type!r}/{text_type!r}"
            )

        layers = _positive(text.get("num_hidden_layers"), "num_hidden_layers")
        mtp_layers = _positive(
            text.get("num_nextn_predict_layers"), "num_nextn_predict_layers"
        )
        ratios = _integers(text.get("compress_ratios"), "compress_ratios")
        if len(ratios) != layers + mtp_layers or any(value not in {0, 1, 2} for value in ratios):
            raise ValueError(
                "DeepSeek-V4.1 compression schedule must cover every backbone and DSpark layer"
            )
        backbone_ratios = ratios[:layers]
        try:
            encoder_layers = backbone_ratios.index(1)
        except ValueError as error:
            raise ValueError("DeepSeek-V4.1 CED schedule has no decoder segment") from error
        if (
            encoder_layers <= 0
            or any(value == 1 for value in backbone_ratios[:encoder_layers])
            or any(value != 1 for value in backbone_ratios[encoder_layers:])
            or any(ratios[layers:])
        ):
            raise ValueError("DeepSeek-V4.1 CED/DSpark compression schedule is inconsistent")

        kv_sources = _integers(text.get("kv_source_layer_ids"), "kv_source_layer_ids")
        index_sources = _integers(
            text.get("index_source_layer_ids"), "index_source_layer_ids"
        )
        candidate_source = int(text.get("candidate_source_layer_id", -1))
        if (
            not kv_sources
            or not set(kv_sources).issubset(index_sources)
            or any(layer < 0 or layer >= layers for layer in index_sources)
            or candidate_source not in index_sources
        ):
            raise ValueError("DeepSeek-V4.1 CSA2 source-layer schedule is inconsistent")
        active_ratio = 0
        kv_source_set = set(kv_sources)
        for layer, ratio in enumerate(backbone_ratios):
            if layer in kv_source_set:
                if ratio <= 0:
                    raise ValueError(
                        "DeepSeek-V4.1 KV source has no compressed stream"
                    )
                active_ratio = ratio
            if ratio > 0 and active_ratio != ratio:
                raise ValueError(
                    "DeepSeek-V4.1 CSA2 consumer has no compatible KV source"
                )

        engram_layers = _integers(text.get("engram_layer_ids"), "engram_layer_ids")
        engram_rows = _integers(
            text.get("engram_num_embeddings"), "engram_num_embeddings"
        )
        if len(engram_layers) != len(engram_rows) or any(
            layer < 0 or layer >= layers for layer in engram_layers
        ):
            raise ValueError("DeepSeek-V4.1 Engram layers and tables disagree")

        dspark_targets = _integers(
            text.get("dspark_target_layer_ids"), "dspark_target_layer_ids"
        )
        if len(dspark_targets) != mtp_layers or any(
            layer < 0 or layer >= layers for layer in dspark_targets
        ):
            raise ValueError("DeepSeek-V4.1 DSpark target layers disagree with stage count")

        hidden = _positive(text.get("hidden_size"), "hidden_size")
        heads = _positive(text.get("num_attention_heads"), "num_attention_heads")
        kv_heads = _positive(text.get("num_key_value_heads"), "num_key_value_heads")
        head_dim = _positive(text.get("head_dim"), "head_dim")
        rope_dim = _positive(text.get("qk_rope_head_dim"), "qk_rope_head_dim")
        groups = _positive(text.get("o_groups"), "o_groups")
        if heads % kv_heads or heads % groups or rope_dim > head_dim or rope_dim % 2:
            raise ValueError("DeepSeek-V4.1 attention dimensions are inconsistent")
        if (
            str(text.get("hidden_act", "")) != "silu"
            or bool(text.get("attention_bias", False))
            or str(text.get("scoring_func", "")) != "sqrtsoftplus"
            or str(text.get("topk_method", "")) != "noaux_tc"
        ):
            raise ValueError("unsupported DeepSeek-V4.1 activation/router semantics")

        quantization = _object(outer.get("quantization_config"), "quantization_config")
        block_size = _integers(
            quantization.get("weight_block_size"), "quantization.weight_block_size"
        )
        if (
            str(quantization.get("quant_method", "")) != "fp8"
            or block_size != (32, 32)
            or str(quantization.get("scale_fmt", "")) != "ue8m0"
            or str(quantization.get("expert_dtype", "")) != "fp4"
        ):
            raise ValueError("unsupported DeepSeek-V4.1 native weight encoding")

        rope = _object(text.get("rope_scaling"), "rope_scaling")
        if str(rope.get("rope_type", "")) != "yarn":
            raise ValueError("DeepSeek-V4.1 requires YaRN RoPE scaling")
        vision = DeepseekV41VisionConfig.from_mapping(
            _object(outer.get("vision_config"), "vision_config")
        )
        return cls(
            family="deepseek_v41",
            model_type=outer_type,
            text_model_type=text_type,
            vocab_size=_positive(text.get("vocab_size"), "vocab_size"),
            hidden_size=hidden,
            moe_intermediate_size=_positive(
                text.get("moe_intermediate_size"), "moe_intermediate_size"
            ),
            num_hidden_layers=layers,
            causal_encoder_layers=encoder_layers,
            num_attention_heads=heads,
            num_key_value_heads=kv_heads,
            head_dim=head_dim,
            qk_rope_head_dim=rope_dim,
            q_lora_rank=_positive(text.get("q_lora_rank"), "q_lora_rank"),
            o_lora_rank=_positive(text.get("o_lora_rank"), "o_lora_rank"),
            o_groups=groups,
            rms_norm_eps=float(text.get("rms_norm_eps", 1e-20)),
            max_position_embeddings=_positive(
                text.get("max_position_embeddings"), "max_position_embeddings"
            ),
            rope_theta=float(text.get("rope_theta", 10_000.0)),
            rope_factor=float(rope.get("factor", 1.0)),
            original_max_position_embeddings=_positive(
                rope.get("original_max_position_embeddings"),
                "rope_scaling.original_max_position_embeddings",
            ),
            n_routed_experts=_positive(text.get("n_routed_experts"), "n_routed_experts"),
            n_shared_experts=_positive(text.get("n_shared_experts"), "n_shared_experts"),
            num_experts_per_tok=_positive(
                text.get("num_experts_per_tok"), "num_experts_per_tok"
            ),
            scoring_func=str(text.get("scoring_func")),
            norm_topk_prob=bool(text.get("norm_topk_prob", False)),
            routed_scaling_factor=float(text.get("routed_scaling_factor", 1.0)),
            swiglu_limit=float(text.get("swiglu_limit", 0.0)),
            sliding_window=_positive(text.get("sliding_window"), "sliding_window"),
            compress_ratios=ratios,
            compress_rope_theta=float(text.get("compress_rope_theta", 160_000.0)),
            kv_source_layer_ids=kv_sources,
            index_source_layer_ids=index_sources,
            index_n_heads=_positive(text.get("index_n_heads"), "index_n_heads"),
            index_head_dim=_positive(text.get("index_head_dim"), "index_head_dim"),
            index_topk=_positive(text.get("index_topk"), "index_topk"),
            candidate_source_layer_id=candidate_source,
            candidate_topk_blocks=_positive(
                text.get("candidate_topk_blocks"), "candidate_topk_blocks"
            ),
            candidate_block_size=_positive(
                text.get("candidate_block_size"), "candidate_block_size"
            ),
            hc_mult=_positive(text.get("hc_mult"), "hc_mult"),
            hc_sinkhorn_iters=_positive(
                text.get("hc_sinkhorn_iters"), "hc_sinkhorn_iters"
            ),
            hc_eps=float(text.get("hc_eps", 1e-6)),
            engram_layer_ids=engram_layers,
            engram_num_embeddings=engram_rows,
            engram_max_ngram_size=_positive(
                text.get("engram_max_ngram_size"), "engram_max_ngram_size"
            ),
            engram_vocab_size=_positive(text.get("engram_vocab_size"), "engram_vocab_size"),
            engram_n_heads=_positive(text.get("engram_n_heads"), "engram_n_heads"),
            engram_head_dim=_positive(text.get("engram_head_dim"), "engram_head_dim"),
            engram_pad_token_id=int(text.get("engram_pad_token_id", 0)),
            engram_compressed_vocab_size=_positive(
                text.get("engram_compressed_vocab_size"), "engram_compressed_vocab_size"
            ),
            num_nextn_predict_layers=mtp_layers,
            dspark_block_size=_positive(text.get("dspark_block_size"), "dspark_block_size"),
            dspark_noise_token_id=int(text.get("dspark_noise_token_id", 0)),
            dspark_target_layer_ids=dspark_targets,
            dspark_markov_rank=_positive(
                text.get("dspark_markov_rank"), "dspark_markov_rank"
            ),
            dspark_n_routed_experts=_positive(
                text.get("dspark_n_routed_experts"), "dspark_n_routed_experts"
            ),
            dspark_num_experts_per_tok=_positive(
                text.get("dspark_num_experts_per_tok"), "dspark_num_experts_per_tok"
            ),
            dense_weight_block_size=(block_size[0], block_size[1]),
            dense_scale_format=str(quantization.get("scale_fmt")),
            expert_dtype=str(quantization.get("expert_dtype")),
            image_token_id=int(outer.get("image_token_id", -1)),
            eos_token_ids=_eos_ids(outer.get("eos_token_id")),
            vision=vision,
        )


def parse_deepseek_v41_config(config: Mapping[str, object]) -> DeepseekV41Config:
    return DeepseekV41Config.from_hf_config(config)


__all__ = [
    "DeepseekV41Config",
    "DeepseekV41VisionConfig",
    "parse_deepseek_v41_config",
]
